lib/hiera/interpolate.rb

```diff
@@ -59,9 +64,19 @@ def do_interpolation(data, scope, extra_data, context)
     end
     private :do_interpolation
 
-    def get_interpolation_method_and_key(data)
+    def get_interpolation_method_and_key(data, is_config)
       if (match = data.match(METHOD_INTERPOLATION))
-        case match[1]
+        key = match[1]
+        if(is_config)
+          if key == 'hiera'
+            # Letting this pass while parsing the config is very likely to result in another attempt to parse
+            # the config and hence, endless recursion.
+            raise Hiera::InterpolationInvalidValue, "Cannot use interpolation method 'hiera' in hiera configuration file"
```

Comment:
This is a breaking change for those than managed to configure their hiera carefully to make it work. This will require them to make changes. To make it an error we need a major HI release, which we do not want. Suggest just making it be a deprecation warning.